README.md
[![Build Status](https://travis-ci.org/LucianoGanga/simple-headless-chrome.svg?branch=master)](https://travis-ci.org/LucianoGanga/simple-headless-chrome)

# simple-headless-chrome

## Important version >= 3.3.0

Version 3.3.0 includes a new feature that allows managing browser tabs. 

This new feature comes with some breaking changes that will allow us future scalability. 

To avoid problems for people that uses version >= 3.3.0 of this module, we supported those breaking changes with methods that will be deprecated in version 4.0.0. 

# Introduction

This is an abstraction to use a Headless version of Google Chrome in a very simple way.
I was inspired by the next projects:

-   Doffy (<https://github.com/qieguo2016/doffy>)
-   Horseman (<https://github.com/johntitus/node-horseman>)
-   chrome-remote-interface (<https://github.com/cyrus-and/chrome-remote-interface>)
-   lighthouse (<https://github.com/googlechrome/lighthouse>)

And I had to read a lot here too:

-   <https://developers.google.com/web/updates/2017/04/headless-chrome>
-   <https://chromedevtools.github.io/devtools-protocol>

And you can also use this in heroku thanks to <https://github.com/heroku/heroku-buildpack-google-chrome>

I built this basically because I got tired of an error I received in an edge case when using PhantomJS (Unhandled reject Error: Failed to load url). So I decided to make my own abstraction, to be used in a heroku app, and simple to use as Horseman.

I didn't have time to document here in the readme, but every method in the source code is documented.

It's really simple to use. I hope I can get some time to make a QuickStart guide + document the API methods here.

You can read my post in Medium about this module: [How to tell to a headless Google Chrome to write a post in Medium for you](https://medium.com/@lucho_1312/how-to-tell-to-a-headless-google-chrome-to-write-a-post-in-medium-for-you-cbee3272f2fa)

# Collaboration

If you want to collaborate with the project, in any way (documentation, examples, fixes, etc), just send a PR :)

If you rock at making tests, it would be very useful if you can help us making this module better.
It's not necesary to build all the tests, but if someone knows how to code the base to add tests to this module, it would really help for someone else to start with this part.

Thank you to everyone who already help submitting a PR! :D

# Installation

## 1) Install Google Chrome Headless

### In your PC

Mac: Chrome Headless is shipped in Chrome Canary. You can install it here: <https://www.google.com/chrome/browser/canary.html>

Linux: Chrome headless is shipped on chrome 59. so you can install Chrome 59 to use the headless mode:

<https://askubuntu.com/questions/79280/how-to-install-chrome-browser-properly-via-command-line>

    sudo apt-get install libxss1 libappindicator1 libindicator7
    wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
    sudo dpkg -i google-chrome*.deb  # Might show "errors", fixed by next line
    sudo apt-get install -f

### In a NodeJS Heroku App

Just add the buildpack for Heroku and vualá! Everything is ready
You can check the buildpack repository here: <https://github.com/heroku/heroku-buildpack-google-chrome>

### Using a Docker image

With the addition of Chrome Remote Interface into Chrome 59, a simple way to install is using the Docker image for Chrome Headless, such as <https://hub.docker.com/r/justinribeiro/chrome-headless/> or <https://hub.docker.com/r/yukinying/chrome-headless/>

If using Docker, in your app, configure for headless as follows:

```js
const browser = new HeadlessChrome({
  headless: true,
  launchChrome: false,
  chrome: {
    host: 'localhost',
    port: 9222, // Chrome Docker default port
    remote: true,
  },
  browserlog: true
})
```

## 2) Install the NPM Module

    npm install --save simple-headless-chrome

# Compatibility

Thanks to @lewisf, simple-headless-chrome is compatible on NodeJS >= 4! 
I hope more persons can benefit of this now :) 

# Usage

```js
const HeadlessChrome = require('simple-headless-chrome')

const browser = new HeadlessChrome({
  headless: true, // If you turn this off, you can actually see the browser navigate with your instructions,
  chrome: {
    userDataDir: '/tmp/headlessDataDir' // This can be null, so a tmp folder will be created and then destroyed
  }
})
```

Once you have the browser instance, you can call the methods to interact with it.

# Methods

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

## inject

Injects a module in the page

**Parameters**

-   `name` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** Name of the module to inject

## evaluate

Evaluates a fn in the context of the browser

**Parameters**

-   `fn`  {function} - The function to evaluate in the browser
-   `args` **...any** {\*} - The arguments to pass to the function

## evaluateAsync

Evaluates an async fn in the context of the browser

**Parameters**

-   `fn`  {function} - The function to evaluate in the browser
-   `args` **...any** {\*} - The arguments to pass to the function

## evaluateOnNode

Evaluates a fn in the context of a passed node

**Parameters**

-   `node` **NodeObject** The Node Object used to get the context
-   `fn`  {function} - The function to evaluate in the browser
-   `args` **...any** {\*} - The arguments to pass to the function

## goTo

Navigates to a URL

**Parameters**

-   `url` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The URL to navigate to
-   `opt`   (optional, default `{}`)
-   `options` **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)** The options object.
    options:

**Properties**

-   `timeout` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Time in ms that this method has to wait until the
    "pageLoaded" event is triggered. If the value is 0 or false, it means that it doesn't
    have to wait after calling the "Page.navigate" method

## getNodeValue

Get the value of an Node.

**Parameters**

-   `node` **NodeObject** The Node Object

Returns **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)** Object containing type and value of the element

## getValue

Get the value of an element.

**Parameters**

-   `selector` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The target selector
-   `frameId` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The FrameID where the selector should be searched

Returns **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)** Object containing type and value of the element

## setNodeValue

Set the value of an element.

**Parameters**

-   `node` **NodeObject** The Node Object
-   `value` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The value to set the node to (it may be an array of values when the node is a multiple "HTMLSelectElement")

## setValue

Set the value of an element.

**Parameters**

-   `selector` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The selector to set the value of.
-   `value` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)?** The value to set the selector to
-   `frameId` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The FrameID where the selector should be searched

## fill

Fills a selector of an input or textarea element with the passed value

**Parameters**

-   `selector` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The selector
-   `value` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The value to fill the element matched in the selector
-   `frameId` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The FrameID where the selector should be searched

## clear

Clear an input field.

**Parameters**

-   `selector` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The selector to clear.
-   `frameId` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The FrameID where the selector should be searched

## querySelector

Returns the node associated to the passed selector

**Parameters**

-   `selector` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The selector to find
-   `frameId` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The FrameID where the selector should be searched

## focus

Focus on an element matching the selector

**Parameters**

-   `selector` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The selector to find the element
-   `frameId` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The FrameID where the selector should be searched

## type

Simulate a keypress on a selector

**Parameters**

-   `selector` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The selector to type into.
-   `text` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The text to type.
-   `frameId` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The FrameID where the selector should be searched
-   `opts`  
-   `options` **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)** Lets you send keys like control & shift

## typeText

Types text (doesn't matter where it is)

**Parameters**

-   `text` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The text to type.
-   `opts`  
-   `options` **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)** Lets you send keys like control & shift

## select

Select a value in an html select element.

**Parameters**

-   `selector` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The identifier for the select element.
-   `value` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The value to select.
-   `frameId` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The FrameID where the selector should be searched

## keyboardEvent

-   **See: <https://chromedevtools.github.io/devtools-protocol/tot/Input/#method-dispatchKeyEvent>**

Fire a key event.

**Parameters**

-   `type` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The type of key event. (optional, default `keypress`)
-   `key` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The key to use for the event. (optional, default `null`)
-   `modifier` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** The keyboard modifier to use. (optional, default `0`)
-   `windowsVirtualKeyCode`   (optional, default `0`)

## wait

Waits certain amount of ms

**Parameters**

-   `time` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Ammount of ms to wait

## onConsole

Binding callback to handle console messages

**Parameters**

-   `listener`  is a callback for handling console message

## waitForPageToLoad

Waits for a page to finish loading. Throws error after timeout

**Parameters**

-   `timeout` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** The timeout in ms. (Default: "loadPageTimeout" property in the browser instance options)

## waitForFrameToLoad

Waits for all the frames in the page to finish loading. Returns the list of frames after that

**Parameters**

-   `url` **([regexp](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp) \| [string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String))** The URL that must be waited for load
-   `timeout`  

Returns **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)** List of frames, with childFrames

## waitForSelectorToLoad

Waits for a selector to finish loading. Throws error after timeout

**Parameters**

-   `selector` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The identifier for the select element.
-   `interval` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** The interval in ms. (Default: "loadPageTimeout" property in the browser instance options)
-   `timeout` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** The timeout in ms. (Default: "loadPageTimeout" property in the browser instance options)

## mouseEvent

-   **See: <https://chromedevtools.github.io/devtools-protocol/tot/Input/#method-dispatchMouseEvent>**

Fire a mouse event.

**Parameters**

-   `$0` **[Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)** 
    -   `$0.type`   (optional, default `'mousePressed'`)
    -   `$0.x`   (optional, default `0`)
    -   `$0.y`   (optional, default `0`)
    -   `$0.modifiers`   (optional, default `0`)
    -   `$0.button`   (optional, default `'left'`)
    -   `$0.clickCount`   (optional, default `1`)
-   `type` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** Type of the mouse event. Allowed values: mousePressed, mouseReleased, mouseMoved. (optional, default `mousePressed`)
-   `x` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** X coordinate of the event relative to the main frame's viewport. (optional, default `0`)
-   `y` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Y coordinate of the event relative to the main frame's viewport. 0 refers to the top of the viewport and Y increases as it proceeds towards the bottom of the viewport. (optional, default `0`)
-   `modifier` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Bit field representing pressed modifier keys. Alt=1, Ctrl=2, Meta/Command=4, Shift=8 (default: 0). (optional, default `0`)
-   `button` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** Mouse button (default: "none"). Allowed values: none, left, middle, right. (optional, default `left`)

## click

Click on a selector by firing a 'click event' directly in the element of the selector

**Parameters**

-   `selector` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** Selector of the element to click
-   `frameId` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The FrameID where the selector should be searched

## clickOnSelector

Clicks left button hover the centroid of the element matching the passed selector

**Parameters**

-   `selector` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)?** 
-   `frameId` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The FrameID where the selector should be searched

## getNodeCentroid

Calculates the centroid of a node by using the boxModel data of the element

**Parameters**

-   `nodeId` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The Node Id

Returns **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)** { x, y } object with the coordinates

## getCookies

Get the browser cookies

Returns **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)** Object with all the cookies

## setCookie

Set the browser cookies

**Parameters**

-   `url` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The request-URI to associate with the setting of the cookie.
-   `name` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The name of the cookie.
-   `value` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The value of the cookie.

Returns **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** True if successfully set cookie

## exist

Checks if an element matches the selector

**Parameters**

-   `selector` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The selector string
-   `frameId` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The FrameID where the selector should be searched

Returns **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** Boolean indicating if element of selector exists or not

## visible

Checks if an element matching a selector is visible

**Parameters**

-   `selector` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The selector string
-   `frameId` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** The FrameID where the selector should be searched

Returns **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** Boolean indicating if element of selector is visible or not

## printToPDF

Prints the page to PDF

**Parameters**

-   `options` **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)** Options object
    Options properties: (optional, default `{}`)
-   `returnBinary` **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** If true, returns as binary. Otherwise, returns a base64 string (optional, default `false`)

**Properties**

-   `landscape` **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** Paper orientation. Defaults to false.
-   `displayHeaderFooter` **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** Display header and footer. Defaults to false.
-   `printBackground` **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** Print background graphics. Defaults to false.
-   `scale` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Scale of the webpage rendering. Defaults to 1.
-   `paperWidth` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Paper width in inches. Defaults to 8.5 inches.
-   `paperHeight` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Paper height in inches. Defaults to 11 inches.
-   `marginTop` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Top margin in inches. Defaults to 1cm (~0.4 inches).
-   `marginBottom` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Bottom margin in inches. Defaults to 1cm (~0.4 inches).
-   `marginLeft` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Left margin in inches. Defaults to 1cm (~0.4 inches).
-   `marginRight` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Right margin in inches. Defaults to 1cm (~0.4 inches).
-   `pageRanges` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** Paper ranges to print, e.g., '1-5, 8, 11-13'. Defaults to the empty string, which means print all pages.

Returns **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** Binary or Base64 string with the PDF data

## getScreenshot

Takes a screenshot of the page and returns it as a string

**Parameters**

-   `captureOptions` **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)** Options object
    Options properties:
    -   `captureOptions.format`   (optional, default `'png'`)
    -   `captureOptions.quality`  
    -   `captureOptions.clip`   (optional, default `{x:0,y:0,width:this.options.deviceMetrics.width,height:this.options.deviceMetrics.height,scale:this.options.deviceMetrics.deviceScaleFactor}`)
    -   `captureOptions.fromSurface`  
    -   `captureOptions.selector`  
    -   `captureOptions.fullPage`  

release-notes.md
# Release notes for version 4.0.0
- Removed breaking changes support for previous versions. 

Now the new syntax is like this: 
```js
const HeadlessChrome = require('simple-headless-chrome')

// This creates a new browser instance
const browser = new HeadlessChrome({
  headless: true 
})
async function navigateWebsite() {
  // This initializates the browser
  await browser.init()

  // This creates and connects to a new tab in the browser
  const mainTab = await browser.newTab({  
    privateTab: false // (Optional) Defaults to false
  })

  // Navigate to a URL
  await mainTab.goTo('http://www.mywebsite.com/login')

  // ... etc ... 

  // Closes the browser
  await browser.close()
```


# Release notes for version 3.3.0
- Add support for tabs and private tabs (last ones only in headless mode)
- Some configuration variables have been moved inside different object properties, to keep certain order and clarity for future features

# IMPORTANT: migration guide to prepare to version 4.0.0

Version 3.3.0 includes support for tabs and also modifies a few configuration variables for initialization. Read this next steps to prepare your code for migrating from versions =< 3.3.* to version > 4.0.0. 

Although v3.3.0 already implements those changes, it keeps support for people that already has their code ready for versions < 4.0.0. 

The next steps include the changes that you need to implement in order to support the next v4.0.0

#### Module initialization Default tab initialization will be deprecated @v4.0.0 of this module. 
Since version 3.3.0 includes tabs, we needed to add a new layer between the navigation and the browser. That new layer allows tabs managing, including incognito tabs too. 
With that new feature we had to change the main syntax of the module, to support more future features in a tidy and scalable way. 
The new syntax involves one extra step: 

Before: 
```js 
const HeadlessChrome = require('simple-headless-chrome')

const browser = new HeadlessChrome({
  headless: true 
})
async function navigateWebsite() {
  await browser.init()
  // Navigate to a URL
  await browser.goTo('http://www.mywebsite.com/login')

  // ...

  // Close the browser
  await browser.close()
}

navigateWebsite()

```

New syntax:
```js
const HeadlessChrome = require('simple-headless-chrome')

const browser = new HeadlessChrome({
  headless: true 
})
async function navigateWebsite() {
  await browser.init({
    startNewFirstTab: false // Only if you use a version > 3.3.0. If you use >= 4.0.0, you don't need to add this setting
  })

  const mainTab = await browser.newTab({  
    privateTab: false // Optional. Defaults to false
  })

  // Navigate to a URL
  await mainTab.goTo('http://www.mywebsite.com/login')

  // ...

  // Close the browser
  await browser.close()
}

navigateWebsite()
``` 

#### Property "disableGPU" was moved inside the "chrome" object. So you need to change this to support the new version

Before: 
```js
const browser = new HeadlessChrome({
  headless: true,
  disableGPU: true
})
``` 

Since version 3.3.0: 
```js
const browser = new HeadlessChrome({
  headless: true,
  chrome: {
    disableGPU: true
  }
})
``` 

#### Properties "loadPageTimeout", "loadSelectorInterval", "loadSelectorTimeout" and "browserLog" were moved inside the "browser" object in the options. 
Before: 
```js
const browser = new HeadlessChrome({
  headless: true,
  loadPageTimeout: 60000,
  loadSelectorInterval: 500,
  loadSelectorTimeout: 60000,
  browserLog: false // {boolean|function}
})
``` 

Since version 3.3.0: 
```js
const browser = new HeadlessChrome({
  headless: true,
  browser: {
    loadPageTimeout: 60000,
    loadSelectorInterval: 500,
    loadSelectorTimeout: 60000,
    browserLog: false // {boolean|function}
  }
})
``` 

lib/browser_.js
'use strict'

const debug = require('debug')('HeadlessChrome:browser')

/**
 * Created by: Luciano A. Ganga
 * Inspired by:
 *    - Doffy (https://github.com/qieguo2016/doffy)
 *    - Horseman (https://github.com/johntitus/node-horseman)
 * More Info:
 * https://developers.google.com/web/updates/2017/04/headless-chrome
 * https://chromedevtools.github.io/devtools-protocol
 * https://github.com/cyrus-and/chrome-remote-interface
 */

const EventEmitter = require('events')

const merge = require('deepmerge')

const chrome = require('./chrome')
const Tab = require('./Tab')

const setupHandlers = require('./setupHandlers')
const setupViewport = require('./setupViewport')
const setupActions = require('./setupActions')

const defaultOptions = {
  headless: true,
  launchChrome: true, // If false, only connects CDP to the Chrome instance in host:port
  chrome: {
    host: 'localhost',
    port: 0, // Chrome instance port (defaults to 0 to generate a random port)
    remote: false,
    launchAttempts: 3,
    userDataDir: null,
    disableGPU: true,
    handleSIGINT: true,
    flags: ['--enable-logging'],
    noSandbox: false
  },
  tabs: {
    maxTabs: Infinity,
    startingUrl: 'about:blank' // Default starting URL for all the tabs
  },
  deviceMetrics: {
    width: 1024,
    height: 768,
    deviceScaleFactor: 0,
    mobile: false,
    fitWindow: false
  },
  browser: {
    loadPageTimeout: 60000,
    loadSelectorInterval: 500,
    loadSelectorTimeout: 60000,
    browserLog: false
  }
}

class HeadlessChrome extends EventEmitter {
  constructor (options = {}) {
    super()
    // Deep merge in order to prevent total override of default params
    this.options = merge(defaultOptions, options)
    this.host = this.options.chrome.host
    this._tabs = [] // TODO: This array must be initialized with the list of tabs of the attached Chrome instance (if the chrome instance is not new but a remote one with existing tabs, this should be autocompleted from that instance)
    this._closedTabs = [] // This keeps track of closed tab ids

    // Add warning and support for old options syntax
    if (typeof this.options.disableGPU !== 'undefined') {
      console.warn(`[simple-headless-chrome] Parameter "disableGPU" is now inside the "chrome" property of the options object. This will be moved in the next version of simple-headless-chrome. + info: https://github.com/LucianoGanga/simple-headless-chrome/blob/master/release-notes.md`)
      this.options.chrome.disableGPU = this.options.disableGPU
    }
    if (typeof this.options.loadPageTimeout !== 'undefined') {
      console.warn(`[simple-headless-chrome] Parameter "loadPageTimeout" is now inside the "browser" property of the options object. This will be moved in the next version of simple-headless-chrome. + info: https://github.com/LucianoGanga/simple-headless-chrome/blob/master/release-notes.md`)
      this.options.browser.loadPageTimeout = this.options.loadPageTimeout
    }
    if (typeof this.options.loadSelectorInterval !== 'undefined') {
      console.warn(`[simple-headless-chrome] Parameter "loadSelectorInterval" is now inside the "browser" property of the options object. This will be moved in the next version of simple-headless-chrome. + info: https://github.com/LucianoGanga/simple-headless-chrome/blob/master/release-notes.md`)
      this.options.browser.loadSelectorInterval = this.options.loadSelectorInterval
    }
    if (typeof this.options.loadSelectorTimeout !== 'undefined') {
      console.warn(`[simple-headless-chrome] Parameter "loadSelectorTimeout" is now inside the "browser" property of the options object. This will be moved in the next version of simple-headless-chrome. + info: https://github.com/LucianoGanga/simple-headless-chrome/blob/master/release-notes.md`)
      this.options.browser.loadSelectorTimeout = this.options.loadSelectorTimeout
    }
    if (typeof this.options.browserLog !== 'undefined') {
      console.warn(`[simple-headless-chrome] Parameter "browserLog" is now inside the "browser" property of the options object. This will be moved in the next version of simple-headless-chrome. + info: https://github.com/LucianoGanga/simple-headless-chrome/blob/master/release-notes.md`)
      this.options.browser.browserLog = this.options.browserLog
    }
  }

  /**
   * Initializates a Chrome instance and opens a new tab
   * If it exceeds the maxTabs defined by user, returns undefined
   * @param {boolean} private Indicates a private tab (supported only in Headless mode)
   * @return - A browser tab instance, with all the actions
   */
  async init ({ startNewFirstTab = true } = {}) {
    if (this.options.launchChrome) {
      // Launch a new Chrome instance and update the instance port
      const initializationOptions = this.options.chrome
      this.chromeInstance = await chrome.launch(this.options.headless, initializationOptions)
      this.port = this.chromeInstance.port
    }

    try {
      // Attach the Chrome Debugging Protocol to the chrome instance
      this.client = await chrome.attachCdpToBrowser(this.host, this.port, this.options.chrome.remote)

      /**
       * Create the first browser tab and append all the actions to the main class
       * NOTE: this is only to support versions =< 3.2.0 of this module. This will
       * be deprecated in a future version.
       * Check https://github.com/LucianoGanga/simple-headless-chrome/blob/master/release-notes.md
       * for more information about this
       */
      if (startNewFirstTab) {
        // Setup Actions, Handlers and Viewport
        debug(`:: init => Preparing Actions, Handlers and Viewport for main tab...`)
        setupHandlers.call(this)
        await setupViewport.call(this)
        setupActions.call(this)
        debug(`:: init => Actions, Handlers and Viewport for main tab prepared successfully!`)
      }

      return this
    } catch (err) {
      await this.chromeInstance.kill()
      throw err
    }
  }

  /**
   * Initializates a new Chrome tab and controls it.
   * If it exceeds the maxTabs defined by user, returns undefined
   * @param {boolean} privateTab Indicates a private tab (supported only in Headless mode)
   * @return - A browser tab instance, with all the actions
   */
  async newTab ({ privateTab = false } = {}) {
    debug(`:: newTab => Opening a new tab in browser. Private tab: ${privateTab}`)

    /**
     * Don't open tab if maxTabs reached, and not open private tab if not headless
     */
    if (Object.keys(this._tabs).length >= this.options.tabs.maxTabs) {
      const error = new Error(`Can't open new tab. Max tabs limit reached`)
      error.maxTabs = this.options.tabs.maxTabs
      error.openedTabs = Object.keys(this._tabs).length
      throw error
    }

    // Create and initialize the new tab
    const tab = new Tab(this.host, this.port, {
      privateTab,
      startingUrl: this.options.tabs.startingUrl
    }, this)

    await tab.init()

    this._tabs[tab.getId()] = tab

    return tab
  }

  getTab (tabId) {
    debug(`:: getTab => Getting tab ${tabId}`)
    return this._tabs[tabId]
  }

  /**
   * Lists all the open tabs ids
   */
  async listTabs () {
    debug(`:: listTabs => Listing all tabs`)
    let array = []
    for (let key in this._tabs) {
      array.push(this._tabs[key].id)
    }
    return array
  }

  /**
   * Lists all the private targets/tabs ids
   * only in headless mode
   */
  listPrivateTabs () {
    debug(`:: listPrivateTabs => Listing all private tabs...`)
    if (!this.options.headless) {
      debug(`:: listPrivateTabs => Can't list private tabs without headless mode!`)
      return
    }

    let array = []
    for (let key in this._tabs) {
      if (this._tabs[key].privateTab) array.push(this._tabs[key].id)
    }
    debug(`:: listPrivateTabs => "${array.length}" private tabs found.`)
    return array
  }

  /**
   * Exit the Chrome instance
   * @param {boolean} exit - Indicates if chromeInstance must be killed with SIGHUP
   * @return {boolean} - true if exited successfully
   */
  async close (exit) {
    debug(`:: close => Closing the browser.`)
    if (!this.client) {
      debug(`:: close => Can not close the browser. Browser instance doesn't exists.`)
      return
    }

    try {
      await this.client.close()
      this.client = null
      this._tabs = []
      this._closedTabs = []
      exit && await this.chromeInstance.kill('SIGHUP')
      debug(`:: close => Browser closed correctly!`)
      return true
    } catch (err) {
      throw err
    }
  }

  /**
   * PRIVATE METHOD - Erases all the information of a closed target/tab.
   * @param {string} the target/tab ID that was closed
   */
  async _closeTarget (tabId) {
    const target = this._browser._tabs[tabId]

    // Save information about the closed tab
    this._browser._closedTabs.push({
      id: tabId,
      closedAt: target.closedAt,
      instance: target
    })

    // Remove the tab from the active tabs list
    delete this._browser._tabs[tabId]
  }

  /**
   * Tells if the target/tab is initializated
   */
  isInitialized () {
    return !!this.client
  }

  /**
   * Support attaching actions
   * @param {String} name - Method name
   * @param {Function} fn - Method implementation
   */
  addAction (name, fn) {
    if (typeof fn === 'function') {
      this[name] = fn
    }
    return this
  }

  /**
   * Logs a text in the console
   * @param {string} text - The text to log
   */
  log (text) {
    console.warn(`Method log() was removed from the browser instance since v3.3.0. You can call it from the target/tab instance`)
    console.log(`[Tab ID: ${this._targetId}] ${text}`)
    return this
  }

  /**
   * Logs a text in the console, only in debug mode
   * @param {string} text - The text to log
   */
  debugLog (text) {
    console.warn(`Method log() was removed from the browser instance since v3.3.0. You can call it from the target/tab instance`)
    debug(`[Tab ID: ${this._targetId}] ${text}`)
    return this
  }
}

module.exports = HeadlessChrome
